Derived net cams for simultaneous net cams

gabs1234 opened the pull request:
I'm not 100% I didn't beak something here. Some unit tests would be pretty nice in this case.


Derived cameras take the names of shared lib cams you have installed in uca. e.g. mock -> netmock.
Maybe should we add some sort of minimum amount of derived net cams in the build settings ?

gabs1234 committed:
First commit : boilerplate for derived interfaces

gabs1234 committed:
Moved property installation to derived classes

Meson builds as many net cameras as there are shared libs in uca
install folder. Networking functions remain in uca net, remote
properties are handled in derived classes.

gabs1234 committed:
Added meson build for derived classes

Forgot to add in previous commit

gabs1234 committed:
Removed prototype derived class object

gabs1234 committed:
Removed prints

gabs1234 committed:
removed useless properties list

MarcusZuber commented:
Good idea, but unfortunately we do not have the plugins installed on the server running the net-instance.
Do you think there is a chance to build the subclass in runtime? Somehow the net creates its interface also at runtime when it connects to the "remote" camera

gabs1234 commented:
Ok, so indeed having a default or minimum number of derived classes as a build parameter would be necessary. Like this you'll get derived net cams even if you don't have any shared lib camera plugins installed.

I'm not sure I follow why you'd like to build the subclasses in runtime ? Currently, the code for the derived classes is built at build time, and the remote camera properties are instantiated at runtime, just like before, but in seperate derived classes.

(I just hated the idea of having `net-derived-x` classes so I made a meson script that replaces the x with installed camera names)

gabs1234 committed:
Added min amount of derived classes to build

Only meson supports this for now

MarcusZuber commented:
The plugins are usually not installed on the "client" side, so reading libuca*.so does not make sense.
I'm currently experimenting with subclassing the net-camera in the class_init and registering the remote properties only in the sub-classes, but I'm not yet satisfied. But then we would not need compile-time copies of the classes. I will proceed with this in January.

gabs1234 committed:
Is this file useful?

MarcusZuber commented:
I'm currently too busy to proceed with a more elegant solution.

MarcusZuber reviewed (changes requested):
I'm currently too busy for a more elegant solution.

Could you change this, then we can stay with this solution:

- Remove the reading of the installed plugins. The whole point of uca-net is, that I don't have them installed on the client side. Maybe the paramter min_derived_net_cameras should be renamed num_derived_net_cameras or simmilar.
- Add CMake builds
- Write the number of generated netXX classes in a header-file (I have an idea to make it a bit more hidden from the user)
- The net-camera (without a number) is now broken (not linked correctly). Make it also working for compatibility.
- Make the default number of netXX-cameras to 10 (just checked my logs and my record so far was 5 cameras at the same time and I don't want to recompile everything the next time we try something special).

gabs1234 committed:
Bug fixes + CMake support

Fixed:
 - [x] `min_derived_net_cameras` to `num_derived_net_cameras`
 - [x] Support for CMake builds
 - [x] Added `NUM_DERIVED_NET_CAMERAS` to the config file
 - [x] Set default number of derived cameras to 10.
 - [ ] Still cannot instantiate the parent class `UcaNetCamera`...

gabs1234 commented:
Fixed:
 - [x] `min_derived_net_cameras` to `num_derived_net_cameras`
 - [x] Fixed CMake builds
 - [x] Added  `NUM_DERIVED_NET_CAMERAS` to the config files
 - [x] Set default number of derived cameras to 10. 
 - [ ] Still cannot instantiate the parent class `UcaNetCamera`... `camera_plugin_get_type()` is the culprit. The derived classes must define this function. But this creates conflicts if I also define it in the parent class (double definition).

MarcusZuber commented:
My gobject knowledge is a bit rusty, so here is my idea in c++ wording.

If you create a pure virtual uca net base class, from which you inherit the netX and the normal net camera it should work. Then you do not have to overwrite the method.



I cannot overwrite a method but only add new ones in gobject classes?

MarcusZuber committed:
Make generated net-cameras out-of-source

MarcusZuber committed:
Create 'net' and 'netX' from common uca-net-base-camera

MarcusZuber commented:
This works now.
Question: Now we require python even for only a single net-camera (without suffix). Should we keep it like this, or should I generate the net once and add it to the source code?

MarcusZuber committed:
Create 'net' and 'netX' from common uca-net-base-camera

tfarago commented:
Can you paste here a simple use case?

tfarago commented:
Also why do we just count them and not add the suffix of the actual camera behind net?

tfarago commented:
I mean, how am I supposed to know which camera is which?

MarcusZuber commented:
Use case (in concert):

```python
from concert.devices.camera.uca import Camera
net_dimax = await Camera("net0",params=host:port etc)
net_phantom = await Camera("net1", params=host:port, etc)
```

Then I don't have mixed up properties, since each remote-property is registered to a different class.

With https://github.com/ufo-kit/libuca/pull/102 you can just write "net".

MarcusZuber commented:
> Also why do we just count them and not add the suffix of the actual camera behind net?

We do not have the remote plugins installed on the client side.

tfarago commented:
Right. Perhaps one could somehow get them from the server but I know that's additional work. I left a comment in the sibling PR too.

MarcusZuber committed:
Fix cmake build

MarcusZuber committed:
Fix meson build
